scipy/stats/_resampling.py

```diff
@@ -922,37 +947,39 @@ def monte_carlo_test(data, rvs, statistic, *, vectorized=None,
         resamples = [rvs_i(size=(batch_actual, n_observations_i))
                      for rvs_i, n_observations_i in zip(rvs, n_observations)]
         null_distribution.append(statistic(*resamples, axis=-1))
-    null_distribution = np.concatenate(null_distribution)
-    null_distribution = null_distribution.reshape([-1] + [1]*observed.ndim)
+    null_distribution = xp.concat(null_distribution)
+    null_distribution = xp.reshape(null_distribution, [-1] + [1]*observed.ndim)
 
     # relative tolerance for detecting numerically distinct but
     # theoretically equal values in the null distribution
-    eps =  (0 if not np.issubdtype(observed.dtype, np.inexact)
-            else np.finfo(observed.dtype).eps*100)
-    gamma = np.abs(eps * observed)
+    eps =  (0 if not xp.isdtype(observed.dtype, ('real floating'))
+            else xp.finfo(observed.dtype).eps*100)
+    gamma = xp.abs(eps * observed)
 
     def less(null_distribution, observed):
         cmps = null_distribution <= observed + gamma
-        pvalues = (cmps.sum(axis=0) + 1) / (n_resamples + 1)  # see [1]
+        cmps = xp.asarray(cmps, dtype=dtype)
```

Comment:
array-api-strict doesn't like adding booleans or, if if `cmps` is integer dtype, performing `__truediv__`. So convert directly to the result dtype. I think p-values are still coming out as `float64`, though... I'll look into it.

Reply:
Added dtype-maintenance check to `test_axis_dtype`.